homeassistant/components/osoenergy/water_heater.py

```diff
@@ -0,0 +1,255 @@
+
+    return profile
+
+
+class OSOEnergyWaterHeater(OSOEnergyEntity, WaterHeaterEntity):
+    """OSO Energy Water Heater Device."""
+
+    _attr_name = None
+    _attr_operation_list = OPERATION_LIST
+    _attr_supported_features = WaterHeaterEntityFeature.TARGET_TEMPERATURE
+    _attr_temperature_unit = UnitOfTemperature.CELSIUS
+    _attr_translation_key = "saga_heater"
+
+    @property
+    def device_info(self) -> DeviceInfo:
+        """Return device information."""
+        return DeviceInfo(
+            identifiers={(DOMAIN, self.device["device_id"])},
+            manufacturer=MANUFACTURER,
+            model=self.device["device_type"],
+            name=self.device["device_name"],
+        )
+
+    @property
+    def name(self) -> str:
+        """Return the name of the water heater."""
+        return self.device["device_name"]
+
+    @property
+    def available(self) -> bool:
+        """Return if the device is available."""
+        return self.device.get("attributes", {}).get("available", False)
+
+    @property
+    def current_operation(self) -> str:
+        """Return current operation."""
+        return OSO_ENERGY_TO_HASS_STATE[self.device["status"]["current_operation"]]
+
+    @property
+    def current_temperature(self) -> float:
+        """Return the current temperature of the heater."""
+        return self.device.get("attributes", {}).get("current_temperature", 0)
+
+    @property
+    def target_temperature(self) -> float:
+        """Return the temperature we try to reach."""
+        return self.device.get("attributes", {}).get("target_temperature", 0)
+
+    @property
+    def target_temperature_high(self) -> float:
+        """Return the temperature we try to reach."""
+        return self.device.get("attributes", {}).get("target_temperature_high", 0)
+
+    @property
+    def target_temperature_low(self) -> float:
+        """Return the temperature we try to reach."""
+        return self.device.get("attributes", {}).get("target_temperature_low", 0)
+
+    @property
+    def min_temp(self) -> float:
+        """Return the minimum temperature."""
+        return self.device.get("attributes", {}).get("min_temperature", HEATER_MIN_TEMP)
+
+    @property
+    def max_temp(self) -> float:
+        """Return the maximum temperature."""
+        return self.device.get("attributes", {}).get("max_temperature", HEATER_MAX_TEMP)
+
+    @property
+    def extra_state_attributes(self) -> Mapping[str, Any]:
+        """Return the state attributes."""
+        attr = {}
+
+        for attribute, ha_name in EXTRA_HEATER_ATTR.items():
+            value = self.device.get("attributes", {}).get(attribute)
+            final = value
+            if attribute == "profile":
+                final = _convert_profile_to_local(value)
+            elif attribute in ("heater_state", "heater_mode", "optimization_mode"):
+                value_key = f"{value}".lower()
+                final = OSO_ENERGY_TO_HASS_STATE.get(value_key, final)
```

Comment:
Why don't you put this functionality into `EXTRA_HEATER_ATTR`?

Reply:
I've modified the logic to be as generic as possible but I see no reason to move the code itself to the constants file.
If any new attributes are added at a later point of time it should be enough to just configure them and their mappings in the const.py file. Profile is a special case that needs processing due to the timezone of the user. 

Is that fine as a solution? If not can you give an example? - I could not find other integrations that implement logic in their constant files.